Internal #215: Window EXCLUDE Functionality

hawkfish opened the pull request:
Implement the SQL windowing EXCLUDE functionality started by @j-w-moebius. This includes:

* Parser and Serialisation extensions
* Implementations and tests for all accelerators
* Changes to the custom window API

Despite the increased complexity of the calling sequences, performance appears unaffected.

fixes: duckdblabs/duckdb-internal#215

Commit:
add parser support for window EXCLUDE clauses

Commit:
implement window EXCLUDE for aggregate functions

Commit:
implement window EXCLUDE for first_value, last_value, nth_value

Commit:
add tests for window EXCLUDE

Commit:
add (de-)serialization of window exclusion clauses

Commit:
increment storage version

Commit:
fix bug related to R API and window EXCLUDE

Commit:
generate serialized_parsed_expression.cpp correctly

Commit:
fix parameter case style

Commit:
undo unintended changes

Commit:
fix serialization mistakes

Commit:
fix leaking EXCLUDE implementation for aggregates

Commit:
add EXCLUDE tests for PARTITION BY without ORDER BY

Commit:
remove unnecessary constructor argument

Commit:
add tests for custom aggregates

Commit:
Merge branch 'master' into temp

Commit:
minor fixes

Commit:
Merge branch 'master' into exclude

Commit:
Merge branch 'partition-order' into exclude

Commit:
Merge branch 'partition-order' into exclude

Commit:
Internal #202: EXCLUDE Result Order

Reverse the expected order of one test
now that we have deterministic ordering of single partitions.

Commit:
Internal #202: Hide Exclude Evaluation

Move the code for WindowSegmentTree::EvaluateInternal to WindowSegmentTreeState.
This hides the implementation details and prepares to fix some memory leaks.

Commit:
Internal #202: Right Memory Leak

Fix the right side memory leak.
Avoid unnecessary per-chunk memory allocation.

Commit:
Internal #202: Custom Window Exclude

Plumb the setting through and throw if it is not supported.

Commit:
Merge branch 'partition-order' into exclude

Commit:
Internal #202: Custom Window Exclude

Change API to pass frame lists instead of window setting.

Commit:
Internal #202: Count Star Exclude

Support the EXCLUDE modifiers for COUNT(*).

Commit:
Internal #202: Constant Window Exclude

Use swap instead of copying.

Commit:
Merge branch 'partition-order' into exclude

Commit:
Merge branch 'main' into partition-order

Commit:
Merge branch 'partition-order' into exclude

Commit:
Internal #202: Windowed LIST Exclude

Implement exclude syntax for array_agg and friends.
Fix non-contiguous NULL detection.

Commit:
Merge branch 'main' into exclude

Commit:
Internal #202: Windowed MODE Exclude

Use a state machine to walk the frames and cover all the deltas.

Commit:
Merge branch 'main' into exclude

Commit:
Internal #215: QUANTILE Frame Refactor

Cleanup up QUANTILE a bit in preparation for composite frames.
This reduces the problem to generalising ReuseIndexes (future work).

Commit:
Merge branch 'main' into exclude

Commit:
Internal #202: Windowed MODE Exclude

Use pointers to existing frames instead of copying them.

Commit:
Internal #215: QUANTILE Reuse Refactor

Convert ReuseIndexes to handle multiple ranges.
Only tested for single range.

Commit:
Merge branch 'main' into exclude

Commit:
Internal #215: QUANTILE Reuse Restriction

Reduce the insert range of Reuse to the current frames
instead of the union of current and previous frames.

Commit:
Internal #215: QUANTILE Exclude Tests

Back out range restriction.
Add median tests.

Commit:
Merge branch 'main' into exclude

Commit:
Internal #202: Windowed Exclude Serialisation

Make the new exclude clause member optional so we don't have to bump the file version.

Commit:
Merge branch 'feature' into exclude

Commit:
Internal #202: Windowed Exclude Serialisation

Still broken but should be closer. Probably just needs a `GetDefault`?

Commit:
Merge branch 'feature' into exclude

Commit:
Internal #202: Windowed Exclude Serialisation

Updated json file with default value. Works now.

Commit:
Internal #215: Deterministic EXCLUDE Tests

Add sorting to handle new work stealing.

Commit:
Internal #358: Exclude API Polishing

Since not all custom window functions need the previous frame,
push responsibility down to the function states that do.
Pass around vectors so we have fewer silly arguments.

Commit:
Internal #358: Frame Intersection Template

Factor out the frame intersection logic for reuse.

Commit:
Internal #358: MODE State Destructor

Fix the ModeState to have proper constructors and destructors
and then call them.

Commit:
Internal #358: MODE Memory Leak

Hack around memory leak in PerfectAggregateHashTable.

taniabogatsch commented on src/execution/window_executor.cpp:
```diff
@@ -602,6 +605,143 @@ void WindowExecutorBoundsState::UpdateBounds(idx_t row_idx, DataChunk &input_chu
+	//! A validity mask consisting of only one entries (needed if no ignore_nulls mask is supplied)
+	ValidityMask all_ones_mask;
+	//! A field of one bits
+	vector<validity_t> one_bits;
```
Why can't this be a `ValidityMask` as well?

j-w-moebius replied:
It's only the raw data used by `all_ones_mask`, not a mask of its own (see l. 625). I remember adding it as an attribute for allocation reasons, but maybe there's another way.

hawkfish replied:
This mask isn't needed at all. `GetValidityEntry` does the right thing if everything is valid.

taniabogatsch commented on src/execution/window_executor.cpp:
```diff
@@ -602,6 +605,143 @@ void WindowExecutorBoundsState::UpdateBounds(idx_t row_idx, DataChunk &input_chu
+};
+
+void ExclusionFilter::FetchFromSource(idx_t begin, idx_t end) {
+	idx_t begin_entry_idx, end_entry_idx, buf;
```
Nit: can we change this to three lines?

taniabogatsch commented on src/execution/window_executor.cpp:
```diff
@@ -602,6 +605,143 @@ void WindowExecutorBoundsState::UpdateBounds(idx_t row_idx, DataChunk &input_chu
+	mask.GetEntryIndex(end, end_entry_idx, buf);
+	auto dst = mask.GetData() + begin_entry_idx;
+	auto src = mask_src->GetData() + begin_entry_idx;
+	for (idx_t entry_idx = begin_entry_idx; entry_idx++ <= end_entry_idx;) {
+		*dst++ = *src++;
+	}
```
Nit: can this be a `memcopy`?

hawkfish replied:
I redid the loop in a different way.

taniabogatsch commented on src/execution/window_executor.cpp:
```diff
@@ -602,6 +605,143 @@ void WindowExecutorBoundsState::UpdateBounds(idx_t row_idx, DataChunk &input_chu
+	}
+}
+
+void ExclusionFilter::ResetMask(idx_t row_idx, idx_t offset) {
```
Is this for reusability in other windows?

hawkfish replied:
No I believe it is for the "Value" executors.

j-w-moebius replied:
> No I believe it is for the "Value" executors.

Exactly.

taniabogatsch commented on src/include/duckdb/common/vector_operations/aggregate_executor.hpp:
```diff
@@ -382,15 +382,74 @@ class AggregateExecutor {
 		}
 	}
 
+	using Frames = vector<FrameBounds>;
```
Nit: can we use `vector<FrameBounds>`?

hawkfish replied:
As I mentioned, I find `using` to be good documentation.

hawkfish replied:
But if @Mytherin disagrees I can certainly change it.

taniabogatsch commented on src/include/duckdb/common/vector_operations/aggregate_executor.hpp:
```diff
@@ -382,15 +382,74 @@ class AggregateExecutor {
+	}
+
+	template <typename OP>
+	static void IntersectFrames(const Frames &lefts, const Frames &rights, OP &op) {
+		const auto union_start = MinValue(rights[0].start, lefts[0].start);
+		const auto union_end = MaxValue(rights.back().end, lefts.back().end);
```
I find it confusing that the function is called `IntersectFrames`, whereas the variable names are `union_...`. From my understanding of this function, we traverse the left frame vector and the right frame vector, and for each value evaluate if it is in neither frame, the left vector, the right vector, or in both vectors. So maybe we can rename them to `min_frame_start` and `max_frame_end`?

hawkfish replied:
Yeah that's not quite right - the proper term is the covering interval.  I'lll fix the names.

taniabogatsch commented on src/include/duckdb/common/vector_operations/aggregate_executor.hpp:
```diff
@@ -382,15 +382,74 @@ class AggregateExecutor {
+		idx_t l = 0;
+		idx_t r = 0;
+		for (auto i = union_start; i < union_end;) {
+			int overlap = 0;
```
Nit: `idx_t`? There are more usages of `int` in this file.

hawkfish replied:
It's a two bit mask. I could make it a `uint8_t`?

taniabogatsch replied:
Sounds good to me!

taniabogatsch commented on src/execution/window_segment_tree.cpp:
I noticed that we pass `WindowExclusion exclude_mode_p` to many functions in this PR and wondered if we should make it `const` within the functions.

taniabogatsch reviewed (changes requested):
Hey @hawkfish! Thanks for the PR, `EXCLUDE` looks like a pretty neat extension to the windowing functionality. I added some comments and nits.

From my understanding, the significant change here is the switch from a single `FrameBounds` parameter to a vector of those? From what I can tell, we then determine what to exclude by setting the validity mask accordingly.

hawkfish commented on src/core_functions/aggregate/holistic/mode.cpp:
I do that for readability... I prefer to use type names that describe functionality/semantics instead of endless nested template descriptions that have to be copied and pasted everywhere and are hard to change.  It's not really any different from using symbolic constants instead of writing the value out everywhere.

hawkfish commented on src/core_functions/aggregate/holistic/mode.cpp:
```diff
@@ -149,7 +144,7 @@ template <typename KEY_TYPE, typename ASSIGN_OP>
 	template <class STATE>
 	static void Initialize(STATE &state) {
-		state.Initialize();
+		new (&state) STATE();
```
Ah it's in `Destroy`. But I can look at using placement `delete` to make it easier to match up.

hawkfish replied:
Looks like [placement `delete` isn't a thing](https://stackoverflow.com/questions/6783993/placement-new-and-delete) and what we have now is correct (especially since the memory is being managed separately.)

hawkfish commented on test/sql/window/test_invalid_window.test:
```diff
@@ -24,3 +24,15 @@ select row_number() over (range between unbounded preceding and unbounded preced
+----
+not implemented
+
+# cannot use a window named "groups"
```
Amusingly this is not actually an error - the query is not syntactically correct!  if you fix the syntax it actually passes.

hawkfish replied:
Yes but its just a syntax error.

hawkfish commented on src/core_functions/aggregate/holistic/quantile.cpp:
```diff
@@ -36,41 +36,75 @@ inline interval_t operator-(const interval_t &lhs, const interval_t &rhs) {
+	}
+
+	inline bool Contains(idx_t i) const {
+		for (size_t f = 0; f < frames.size(); ++f) {
```
I've been wondering about that. The type of `size` is `size_t`. I think of `idx_t` as being something we use to index data larger than memory?

j-w-moebius commented on third_party/libpg_query/grammar/statements/select.y:
```diff
@@ -3080,7 +3080,7 @@ window_specification: '(' opt_existing_window_name opt_partition_clause
 
 /*
- * If we see PARTITION, RANGE, or ROWS as the first token after the '('
+ * If we see PARTITION, RANGE, ROWS or GROUPS as the first token after the '('
```
> Does this also mean that windows cannot be called partition, range, and rows?

Yes. I didn't write tests for them because my changes to the parser only affected `groups`, but they could be added for the sake of completeness.

hawkfish commented on src/core_functions/aggregate/holistic/quantile.cpp:
```diff
@@ -36,41 +36,75 @@ inline interval_t operator-(const interval_t &lhs, const interval_t &rhs) {
 }
 
+struct FrameSet {
+	using FrameVec = vector<FrameBounds>;
```
At the very least it should have the same name as in `AggregateExecutor`...

hawkfish committed:
Internal #215: Window EXCLUDE Feedback

Incorporate review feedback.
Most notably, the ValidityMask object is smart enough to handle
the AllValid situation without constructing extra bitmasks.

taniabogatsch commented:
Thanks for all the comments and explanations on my notes! LGTM now.

- I mentioned the `size_t` stuff because of our C++ Guidelines written down [here](https://github.com/duckdb/duckdb/blob/main/CONTRIBUTING.md)
- As for `using`, it is personal preference, and I understand your reasoning, I was mostly commenting because I hadn't seen it much/anywhere else in our codebase yet! :)

hawkfish committed:
Merge branch 'feature' into exclude

Mytherin commented:
Thanks!
